src/Semantic/Api/Terms.hs

```diff
@@ -96,26 +95,29 @@ quietTerm blob = showTiming blob <$> time' ( asks showTermParsers >>= \ parsers
       in stringUtf8 (status <> "\t" <> show (blobLanguage blob) <> "\t" <> blobPath blob <> "\t" <> show duration <> " ms\n")
 
 
-type ParseEffects sig m = (Member (Error SomeException) sig, Member (Reader PerLanguageModes) sig, Member Parse sig, Member (Reader Config) sig, Carrier sig m)
-
-
 showTermParsers :: PerLanguageModes -> Map Language (SomeParser ShowTerm Loc)
 showTermParsers = allParsers
 
 class ShowTerm term where
   showTerm :: (Carrier sig m, Member (Reader Config) sig) => term Loc -> m Builder
 
-instance (Functor syntax, Show1 syntax) => ShowTerm (Term syntax) where
-  showTerm = serialize Show . quieterm
+instance (TermMode term ~ strategy, ShowTermBy strategy term) => ShowTerm term where
+  showTerm = showTermBy @strategy
+
+class ShowTermBy (strategy :: LanguageMode) term where
+  showTermBy :: (Carrier sig m, Member (Reader Config) sig) => term Loc -> m Builder
+
+instance ShowTermBy 'Precise Java.Term where
+  showTermBy = serialize Show . void . Java.getTerm
 
-instance ShowTerm Java.Term where
-  showTerm = serialize Show . void . Java.getTerm
+instance ShowTermBy 'Precise JSON.Term where
+  showTermBy = serialize Show . void . JSON.getTerm
 
-instance ShowTerm JSON.Term where
-  showTerm = serialize Show . void . JSON.getTerm
+instance ShowTermBy 'Precise PythonPrecise.Term where
+  showTermBy = serialize Show . void . PythonPrecise.getTerm
 
-instance ShowTerm Python.Term where
-  showTerm = serialize Show . void . Python.getTerm
```

Comment:
We could combine these instances too if we had a class abstracting precise terms to their underlying AST types. That’s something for `semantic-ast`, perhaps. (cc @aymannadeem)

Reply:
happy to discuss next steps for `semantic-ast` tomorrow, sorry for the lag in catching up to things

Reply:
No apologies necessary, async is totes fine! 👍